node/src/components/consensus/protocols/highway/participation.rs

```diff
@@ -0,0 +1,96 @@
+use std::cmp::Reverse;
+
+use itertools::Itertools;
+
+use crate::{
+    components::consensus::{
+        highway_core::{
+            highway::Highway,
+            state::{Fault, State},
+            validators::ValidatorIndex,
+        },
+        traits::Context,
+    },
+    types::Timestamp,
+};
+
+/// A validator's participation status: whether they are faulty or inactive.
+#[derive(Copy, Clone, Debug, Ord, PartialOrd, Eq, PartialEq)]
+enum Status {
+    LastSeenSecondsAgo(u64),
+    Inactive,
+    EquivocatedInOtherEra,
+    Equivocated,
+}
+
+impl Status {
+    /// Returns a `Status` for a validator unless they are honest and online.
+    fn for_index<C: Context>(
+        idx: ValidatorIndex,
+        state: &State<C>,
+        now: Timestamp,
+    ) -> Option<Status> {
+        if let Some(fault) = state.maybe_fault(idx) {
+            return Some(match fault {
+                Fault::Banned | Fault::Indirect => Status::EquivocatedInOtherEra,
+                Fault::Direct(_) => Status::Equivocated,
+            });
+        }
+        if state.panorama()[idx].is_none() {
+            return Some(Status::Inactive);
+        }
+        if state.last_seen(idx) + state.params().max_round_length() < now {
+            let seconds = (now - state.last_seen(idx)).millis() / 1000;
+            return Some(Status::LastSeenSecondsAgo(seconds));
+        }
+        None
```

Comment:
So we return `None` if validator is honest but we haven't seen new unit in the last `max_round_length` seconds window?

Reply:
Yes. It's stricter than the "offline" criterion for ping purposes, but I think it makes sense to log someone once the timestamps indicate they've missed a round.